### PR TITLE
Don't allow semicolon in host/service name

### LINKED ISSUE #312: Parse Error when adding a node with semicolon in the varname
## Describe the bug
A clear and concise description of what the issue is.
One of our colleagues added a new node with a name like "vSAN Nuremberg; Citrix Worker" after saving the config the complete business-config fails and is uneditable anymore over the webbrowser only via console with vim / nano to the file in "/etc/icingaweb2/modules/businessprocess/processes"

Error is like 
<code>Parse error on /etc/icingaweb2/modules/businessprocess/processes/node.conf:<row>: No semicolon allowed in varname</code>

Similar behavior as in https://github.com/Icinga/icingaweb2-module-businessprocess/issues/71 and https://github.com/Icinga/icingaweb2-module-businessprocess/issues/26

## Expected behavior
I can think of several options here, best way was to show an error before saving the config file with something like "nasty character semicolon (for example) was used, please check" or some change of parsing so the website and the business config / node is still editable.
Our colleagues do not have access to the businessmodule director to correct it via text editor themselves.

image

## Your Environment
icingaweb2: v 2.9.3
module businessprocess: v 2.3.0
PHP: v 7.3.29
OS: RHEL 7

### PR DESCRIPTION
**Pull request by sukhwinder33445**
fixes #312

### CODE REVIEW

File: application/forms/AddNodeForm.php [hunk 1]
```diff
@@ -87,6 +87,18 @@ protected function addNewProcess()
                 'This is the unique identifier of this process'
             ),
             'validators'    => [
+                [
+                    'validator' => 'Regex',
+                    'options'   => [
+                        'pattern' => '/^[a-zA-Z0-9](?:[a-zA-Z0-9 ._-]*)?[a-zA-Z0-9_]$/',
```
Comment (nilmerg) [inline application/forms/AddNodeForm.php:93]: This allows the underscore also at the end.

File: application/forms/AddNodeForm.php [hunk 2]
```diff
@@ -87,6 +87,18 @@ protected function addNewProcess()
...
+                        'messages'  => [
+                            'regexNotMatch' => $this->translate(
+                                'Id must only consist of alphanumeric characters.
+                                Space, dot, dash and underscore are allowed but not at the beginning or end'
```
Comment (nilmerg) [inline application/forms/AddNodeForm.php:97]: Close parts of multiline strings properly.

e.g.
```
call(
    'foo'
    . ' bar'
);
```

Comment (nilmerg) [reply application/forms/AddNodeForm.php:97]: Adjust this every where else in your changes as well.

File: application/forms/AddNodeForm.php [hunk 3]
```diff
@@ -208,7 +220,18 @@ protected function selectHost()
             'description'  => $this->translate(
                 'Hosts that should be part of this business process node'
             ),
-            'validators'    => [[new NoDuplicateChildrenValidator($this, $this->bp, $this->parent), true]]
+            'validators'    => [
+                [new NoDuplicateChildrenValidator($this, $this->bp, $this->parent), true],
+                ['Callback', true, [
+                    'callback' => function ($value) {
+                        $values = explode(';Hoststatus', $value);
+                        return (bool) preg_match('/^(?:(?!;).)*$/', $values[0]); // object name contains semicolon
```
Comment (nilmerg) [inline application/forms/AddNodeForm.php:228]: `strpos() === false`

File: application/forms/AddNodeForm.php [hunk 4]
```diff
@@ -240,6 +263,19 @@ protected function addHostElement()
             'ignore'       => true,
             'class'        => 'autosubmit',
             'multiOptions' => $this->optionalEnum($this->enumHostForServiceList()),
+            'validators'    => [
+                [
+                    'validator' => 'Regex',
+                    'options'   => [
+                        'pattern' => '/^(?:(?!;).)*$/',
```
Comment (nilmerg) [inline application/forms/AddNodeForm.php:270]: `strpos() === false`

File: application/forms/EditNodeForm.php [hunk 5]
```diff
@@ -223,6 +246,19 @@ protected function addHostElement()
             'ignore'       => true,
             'class'        => 'autosubmit',
             'multiOptions' => $this->optionalEnum($this->enumHostForServiceList()),
+            'validators'    => [
+                [
+                    'validator' => 'Regex',
+                    'options'   => [
+                        'pattern' => '/^(?:(?!;).)*$/',
```
Comment (nilmerg) [inline application/forms/EditNodeForm.php:253]: `strpos() === false`

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: application/forms/AddNodeForm.php [hunk 6]
```diff
@@ -195,7 +207,17 @@ protected function selectHost()
             'description'  => $this->translate(
                 'Hosts that should be part of this business process node'
             ),
-            'validators'    => [[new NoDuplicateChildrenValidator($this, $this->bp, $this->parent), true]]
+            'validators'   => [
+                [new NoDuplicateChildrenValidator($this, $this->bp, $this->parent), true],
+                ['Callback', true, [
+                    'callback' => function ($value) {
+                        return substr_count(';', $value) === 2;
```
[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply application/forms/AddNodeForm.php]: oops, but now it should work properly.

File: application/forms/AddNodeForm.php [hunk 7]
```diff
@@ -88,6 +88,18 @@ protected function addNewProcess()
                 'This is the unique identifier of this process'
             ),
             'validators'    => [
+                [
+                    'validator' => 'Regex',
+                    'options'   => [
+                        'pattern'   => '/^[a-zA-Z0-9](?:[\w\h.-]*)?[a-zA-Z0-9]$/',
```
Comment (nilmerg) [inline application/forms/AddNodeForm.php:94]: This still allows a trailing underscore. Just did a test and hosts/services as well as processes can have a trailing underscore while on master. What's the issue with it?
https://github.com/Icinga/icingaweb2-module-businessprocess/blob/da2ed1aef4848bfa8194f40de423337f9d4373e6/application/forms/BpConfigForm.php#L29

**Pull request closed without merging**